dll/win32/fltlib/fltlib.c

```diff
@@ -109,48 +156,134 @@ FilterLoadUnload(_In_z_ LPCWSTR lpFilterName,
+                                     lpInBuffer,
+                                     nInBufferSize,
+                                     lpOutBuffer,
+                                     nOutBufferSize);
+        }
+        else
+        {
+            Status = NtDeviceIoControlFile(hFltMgr,
+                                           lpOverlapped->hEvent,
+                                           NULL,
+                                           ApcContext,
+                                           (PIO_STATUS_BLOCK)lpOverlapped,
+                                           dwControlCode,
+                                           lpInBuffer,
+                                           nInBufferSize,
+                                           lpOutBuffer,
+                                           nOutBufferSize);
+        }
+
+        /* Return the number of bytes transferred if no error */
+        if (!NT_ERROR(Status) && lpBytesReturned)
+        {
+            *lpBytesReturned = (DWORD)lpOverlapped->InternalHigh;
+        }
+    }
+    else
     {
-        Status = NtWaitForSingleObject(Handle, FALSE, NULL);
-        if (NT_SUCCESS(Status))
+        /* Send file system control or device control according to the device type */
+        if (DeviceType == FILE_DEVICE_FILE_SYSTEM)
+        {
+            Status = NtFsControlFile(hFltMgr,
+                                     NULL,
+                                     NULL,
+                                     NULL,
+                                     &IoStatusBlock,
+                                     dwControlCode,
+                                     lpInBuffer,
+                                     nInBufferSize,
+                                     lpOutBuffer,
+                                     nOutBufferSize);
+        }
+        else
         {
-            Status = IoStatusBlock.Status;
+            Status = NtDeviceIoControlFile(hFltMgr,
+                                           NULL,
+                                           NULL,
+                                           NULL,
+                                           &IoStatusBlock,
+                                           dwControlCode,
+                                           lpInBuffer,
+                                           nInBufferSize,
+                                           lpOutBuffer,
+                                           nOutBufferSize);
+        }
+
+        /* Wait if operation still in progress */
+        if (Status == STATUS_PENDING)
+        {
+            Status = NtWaitForSingleObject(hFltMgr, FALSE, NULL);
+            if (NT_SUCCESS(Status))
+            {
+                Status = IoStatusBlock.Status;
+            }
+        }
+
+        /* Return the number of bytes transferred if no error,
+         * but seems Windows didn't check this. */
+        if (!NT_ERROR(Status))
+        {
+            *lpBytesReturned = (DWORD)IoStatusBlock.Information;
```

Comment:
I understand that whether `Information` is meaningful is based on `NT_ERROR`, but it still seems a bit weird in this function to set bytes returned in some but not all failure cases. Perhaps set it to 0 at the beginning or in the `else` block?

Reply:
Seems Windows neither set it to 0 at the beginning, nor check `if (!NT_ERROR(Status))`...
I guess so, `lpBytesReturned` output is valid only if this function success, so there is no big problem, but a bit weird.

Reply:
Sorry for my misunderstanding, there are some details (may include your question) seems missed by me, more investigations and code fixes are in progress, this PR convert to draft temporarily, my fault

Reply:
I set it to 0 at the beginning and removed `if (!NT_ERROR(Status))`.
`IoStatusBlock.Information` is request-depended and may meaningful when failed, like `STATUS_BUFFER_OVERFLOW` (Warning), `STATUS_FLT_BUFFER_TOO_SMALL` (Warning), `STATUS_BUFFER_TOO_SMALL` (Error), ...